### PR TITLE
Setting sessionId to be same as partition key and adding validation

### PR DESCRIPTION
**Pull request by yvgopal**
To prevent setting partition key to be different from sessionid. ServiceBus requires that SessionId and PartitionKey should be same, if SessionId is set on a message.

### CODE REVIEW

File: sdk/servicebus/Microsoft.Azure.ServiceBus/src/Message.cs [hunk 1]
```diff
@@ -140,6 +146,7 @@ public string SessionId
             {
                 Message.ValidateSessionId(nameof(this.SessionId), value);
                 this.sessionId = value;
+                this.PartitionKey = value;
```
Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.ServiceBus/src/Message.cs:149]: Why not just do the same validation here as opposed to overriding the PartitionKey. It doesn't feel right to silently overwrite it when we throw for the other scenario.

**Comment by JoshLove-msft** [comment]
I'll just note that the service has already implemented this change in behavior, so customers who were setting the session and partition differently would already start seeing errors before this client side validation. The purpose of this client side validation is to avoid an unnecessary service call, and provide a more user-friendly error. Is that correct?

File: sdk/servicebus/Microsoft.Azure.ServiceBus/src/Message.cs [hunk 2]
```diff
@@ -99,6 +99,12 @@ public string PartitionKey
 
             set
             {
+                if (this.sessionId != null && !this.sessionId.Equals(value))
```
Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.ServiceBus/src/Message.cs:102]: Would be more idiomatic to just do sessionId != value.

File: sdk/servicebus/Microsoft.Azure.ServiceBus/src/Message.cs [hunk 3]
```diff
@@ -99,6 +99,12 @@ public string PartitionKey
...
+                {
+                    // SessionId is set. Then partition key must be same as session id.
+                    throw new InvalidOperationException("PartitionKey:" + value + " is not same as SessionId:" + this.sessionId);
```
Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.ServiceBus/src/Message.cs:105]: nit: use string interpolation - $"PartitionKey: {value} is not same as SessionId: {this.sessionId}";

Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.ServiceBus/src/Message.cs:105]: Can we add a test case for this?

**Comment by yvgopal** [comment]
Yes.

---
In reply to: [668915635](https://github.com/Azure/azure-sdk-for-net/pull/13961#issuecomment-668915635) [](ancestors = 668915635)

File: sdk/servicebus/Microsoft.Azure.ServiceBus/src/Message.cs (hunk 1 above)
Comment (yvgopal) [reply sdk/servicebus/Microsoft.Azure.ServiceBus/src/Message.cs:149]: That's how other service bus SDKs do it. it was missing from .net core SDK and I am just adding that missing functionality. SDK treats SessionId as the dominant property and dominant property overwrites the other property.

---
In reply to: [465416026](https://github.com/Azure/azure-sdk-for-net/pull/13961#discussion_r465416026) [](ancestors = 465416026)

**Comment by JoshLove-msft** [comment]
Looks like this change has broken some existing tests:
https://dev.azure.com/azure-sdk/public/_build/results?buildId=484545&view=ms.vss-test-web.build-test-results-tab&runId=12429438&resultId=100303&paneView=debug

**Review by JoshLove-msft — APPROVED** [review]

**Review by shankarsama — APPROVED** [review]
:shipit:

**Pull request merged**